File: packages/sites/src/private-storage.ts
import {
  createCipheriv,
  createDecipheriv,
  createPrivateKey,
  createPublicKey,
  hkdfSync,
  randomBytes,
} from "node:crypto";
import {
  canonical,
  hash,
  validateIdentity,
  type Identity,
} from "../../core/src/index";
import { exactShape } from "../../core/src/protocol";
import {
  RegistryIntegrityError,
  RegistryCapacityError,
  type RegistryTransaction,
} from "../../groups/src/storage";

const DOMAIN = "relayloom/site-private/1";
const RESOURCE_DOMAIN = "relayloom/site-resource-private/1";
type Namespace = "site" | "resource";
export const SITE_PRIVATE_LIMITS = Object.freeze({
  bytes: 6 * 1024 * 1024 + 16384,
  chunkBytes: 512 * 1024,
  chunks: 16,
});
const overhead = 61; // version + salt32 + nonce12 + tag16
function insist(value: unknown, error: string): asserts value {
  if (!value) throw new RegistryIntegrityError(error);
}
function checkedKey(key: string, namespace: Namespace) {
  insist(
    (namespace === "site"
      ? /^site:[a-f0-9]{64}:(record|stage)$/
      : /^resource:[a-f0-9]{64}:(record|stage)$/
    ).test(key),
    "Chave privada de site inválida",
  );
  return key;
}
const part = (key: string, n: number) => key + ":" + String(n).padStart(2, "0");

/** Site preparation contains unpublished signatures. Encrypt it with a key
 * derived from signing ownership, inside the existing transactional encrypted
 * store, so the content-reading secret alone cannot extract those signatures.
 * No existing profile/database format or key derivation is changed. */
export class SitePrivateRecords {
  private closed = false;
  private constructor(
    private tx: RegistryTransaction,
    private ownerId: string,
    private storeId: string,
    private secret: Buffer,
    private namespace: Namespace,
  ) {}
  private check() {
    if (this.closed) throw new Error("Sessão privada de site terminada");
  }
  static run<T>(
    tx: RegistryTransaction,
    identity: Identity,
    fn: (records: SitePrivateRecords) => T,
  ): T {
    return this.session(tx, identity, fn, "site");
  }
  static runResource<T>(
    tx: RegistryTransaction,
    identity: Identity,
    fn: (records: SitePrivateRecords) => T,
  ): T {
    return this.session(tx, identity, fn, "resource");
  }
  private static session<T>(
    tx: RegistryTransaction,
    identity: Identity,
    fn: (records: SitePrivateRecords) => T,
    namespace: Namespace,
  ): T {
    insist(
      namespace === "site" || namespace === "resource",
      "Espaço privado inválido",
    );
    if (
      !validateIdentity(identity.public) ||
      tx.indexBody().owner !== identity.public.id
    )
      throw new Error("Proprietário local inválido");
    let records: SitePrivateRecords | undefined;
    const secret = Buffer.from(identity.signSecret, "base64");
    try {
      const key = createPrivateKey({
        key: secret,
        type: "pkcs8",
        format: "der",
      });
      insist(
        key.asymmetricKeyType === "ed25519" &&
          createPublicKey(key)
            .export({ type: "spki", format: "der" })
            .toString("base64") === identity.public.signKey,
        "A chave privada não pertence ao proprietário",
      );
      records = new SitePrivateRecords(
        tx,
        identity.public.id,
        tx.indexBody().storeId,
        secret,
        namespace,
      );
      const result = fn(records);
      if (result && typeof (result as any).then === "function")
        throw new Error("Operação privada deve ser síncrona");
      return result;
    } catch (error) {
      return tx.abort(
        error instanceof Error ? error : new Error(String(error)),
      );
    } finally {
      if (records) records.closed = true;
      secret.fill(0);
    }
  }
  private get domain() {
    return this.namespace === "site" ? DOMAIN : RESOURCE_DOMAIN;
  }
  private aad(key: string) {
    return Buffer.from(
      canonical([
        this.domain,
        this.ownerId,
        this.storeId,
        checkedKey(key, this.namespace),
      ]),
    );
  }
  private encrypt(key: string, bytes: Buffer) {
    const salt = randomBytes(32),
      nonce = randomBytes(12);
    const derived = Buffer.from(
      hkdfSync("sha256", this.secret, salt, this.domain, 32),
    );
    try {
      const cipher = createCipheriv("aes-256-gcm", derived, nonce);
      cipher.setAAD(this.aad(key));
      const data = Buffer.concat([cipher.update(bytes), cipher.final()]);
      return Buffer.concat([
        Buffer.from([1]),
        salt,
        nonce,
        cipher.getAuthTag(),
        data,
      ]);
    } finally {
      derived.fill(0);
    }
  }
  private decrypt(key: string, bytes: Buffer) {
    insist(
      bytes.length >= overhead && bytes[0] === 1,
      "Envelope privado de site inválido",
    );
    const derived = Buffer.from(
      hkdfSync("sha256", this.secret, bytes.subarray(1, 33), this.domain, 32),
    );
    try {
      const cipher = createDecipheriv(
        "aes-256-gcm",
        derived,
        bytes.subarray(33, 45),
      );
      cipher.setAAD(this.aad(key));
      cipher.setAuthTag(bytes.subarray(45, 61));
      return Buffer.concat([cipher.update(bytes.subarray(61)), cipher.final()]);
    } catch {
      throw new RegistryIntegrityError(
        "Autenticação dos dados privados de site falhou",
      );
    } finally {
      derived.fill(0);
    }
  }
  read(key: string): unknown | null {
    this.check();
    try {
      return this.readChecked(key);
    } catch (error) {
      return this.tx.abort(
        error instanceof Error ? error : new Error(String(error)),
      );
    }
  }
  private readChecked(key: string): unknown | null {
    checkedKey(key, this.namespace);
    const manifest = this.tx.get(key),
      keys = this.tx.keys(key + ":");
    if (!manifest) {
      insist(keys.length === 0, "Blocos privados de site sem índice");
      return null;
    }
    let data: any;
    try {
      data = JSON.parse(manifest.toString("utf8"));
    } catch {
      throw new RegistryIntegrityError("Índice privado de site ilegível");
    }
    insist(
      manifest.length <= 1024 &&
        exactShape(data, ["domain", "bytes", "chunks", "digest"]) &&
        data.domain === this.domain &&
        canonical(data) === manifest.toString("utf8") &&
        Number.isSafeInteger(data.bytes) &&
        data.bytes >= overhead &&
        data.bytes <= SITE_PRIVATE_LIMITS.bytes + overhead &&
        Number.isSafeInteger(data.chunks) &&
        data.chunks ===
          Math.ceil(data.bytes / SITE_PRIVATE_LIMITS.chunkBytes) &&
        data.chunks <= SITE_PRIVATE_LIMITS.chunks &&
        typeof data.digest === "string" &&
        /^[a-f0-9]{64}$/.test(data.digest),
      "Índice privado de site inválido",
    );
    insist(
      keys.length === data.chunks &&
        keys.every((keyPart, i) => keyPart === part(key, i)),
      "Conjunto privado de site incompleto",
    );
    const pieces = keys.map((keyPart, i) => {
      const bytes = this.tx.get(keyPart);
      insist(
        bytes &&
          bytes.length ===
            Math.min(
              SITE_PRIVATE_LIMITS.chunkBytes,
              data.bytes - i * SITE_PRIVATE_LIMITS.chunkBytes,
            ),
        "Bloco privado de site truncado",
      );
      return bytes;
    });
    const encrypted = Buffer.concat(pieces, data.bytes);
    insist(hash(encrypted) === data.digest, "Hash privado de site inválido");
    const plain = this.decrypt(key, encrypted);
    try {
      const value = JSON.parse(plain.toString("utf8"));
      insist(
        Buffer.from(canonical(value)).equals(plain),
        "Dados privados de site não canónicos",
      );
      return value;
    } catch (error) {
      if (error instanceof RegistryIntegrityError) throw error;
      throw new RegistryIntegrityError("Dados privados de site ilegíveis");
    }
  }
  write(key: string, value: unknown) {
    this.check();
    try {
      this.writeChecked(key, value);
    } catch (error) {
      this.tx.abort(error instanceof Error ? error : new Error(String(error)));
    }
  }
  private writeChecked(key: string, value: unknown) {
    checkedKey(key, this.namespace);
    const bytes = Buffer.from(canonical(value));
    if (bytes.length > SITE_PRIVATE_LIMITS.bytes)
      throw new RegistryCapacityError("Dados privados de site acima do limite");
    const encrypted = this.encrypt(key, bytes),
      chunks = Math.ceil(encrypted.length / SITE_PRIVATE_LIMITS.chunkBytes);
    insist(
      chunks <= SITE_PRIVATE_LIMITS.chunks,
      "Demasiados blocos privados de site",
    );
    this.remove(key);
    for (let i = 0; i < chunks; i++)
      this.tx.put(
        part(key, i),
        encrypted.subarray(
          i * SITE_PRIVATE_LIMITS.chunkBytes,
          (i + 1) * SITE_PRIVATE_LIMITS.chunkBytes,
        ),
      );
    this.tx.put(
      key,
      Buffer.from(
        canonical({
          domain: this.domain,
          bytes: encrypted.length,
          chunks,
          digest: hash(encrypted),
        }),
      ),
    );
  }
  /** Explicit removal is for completed/cancelled staging. It is not recovery
   * from a corrupt record; callers must verify the catalog transition first. */
  remove(key: string) {
    this.check();
    checkedKey(key, this.namespace);
    for (const existing of this.tx.keys(key + ":")) this.tx.delete(existing);
    this.tx.delete(key);
  }
}

File: native/sites/private_storage.go
package sites

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/hkdf"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"

	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupstore"
)

const PrivateBytes = 6*1024*1024 + 16384
const PrivateChunkBytes = 512 * 1024
const PrivateChunks = 16
const privateOverhead = 61
const privateDomain = "relayloom/site-private/1"
const resourcePrivateDomain = "relayloom/site-resource-private/1"
const contributionPrivateDomain = "relayloom/site-contribution-private/1"

var privateKeyPattern = regexp.MustCompile(`^site:[a-f0-9]{64}:(record|stage)$`)
var resourcePrivateKeyPattern = regexp.MustCompile(`^resource:[a-f0-9]{64}:(record|stage)$`)
var contributionPrivateKeyPattern = regexp.MustCompile(`^contribution:[a-f0-9]{64}:(record|stage)$`)

type PrivateRecords struct {
	tx             *groupstore.Tx
	owner, storeID string
	secret         []byte
	closed         bool
	namespace      string
}

func privateIntegrity(reason string) error {
	return fmt.Errorf("%w: %s", groupstore.ErrIntegrity, reason)
}

// RunPrivate protects unpublished signatures with signing ownership inside the
// existing transactional profile. It does not change the outer storage format.
func RunPrivate(tx *groupstore.Tx, identity core.Identity, callback func(*PrivateRecords) error) error {
	return runPrivateNamespace(tx, identity, callback, "site")
}
func RunResourcePrivate(tx *groupstore.Tx, identity core.Identity, callback func(*PrivateRecords) error) error {
	return runPrivateNamespace(tx, identity, callback, "resource")
}
func RunContributionPrivate(tx *groupstore.Tx, identity core.Identity, callback func(*PrivateRecords) error) error {
	return runPrivateNamespace(tx, identity, callback, "contribution")
}
func runPrivateNamespace(tx *groupstore.Tx, identity core.Identity, callback func(*PrivateRecords) error, namespace string) error {
	if namespace != "site" && namespace != "resource" && namespace != "contribution" {
		return tx.Abort(privateIntegrity("espaço privado inválido"))
	}
	owner, err := tx.Owner()
	if err != nil {
		return err
	}
	if err = core.ValidateIdentity(identity.Public); err != nil || owner != identity.Public.ID {
		return errors.New("proprietário privado de site inválido")
	}
	if _, err = core.SignCertificateData(identity, []byte("relayloom/site-private-key-check/1")); err != nil {
		return tx.Abort(err)
	}
	secret, err := base64.StdEncoding.DecodeString(identity.SignSecret)
	if err != nil {
		return tx.Abort(err)
	}
	storeID, err := tx.StoreID()
	if err != nil {
		clear(secret)
		return err
	}
	records := &PrivateRecords{tx: tx, owner: owner, storeID: storeID, secret: secret, namespace: namespace}
	defer func() { records.closed = true; clear(records.secret) }()
	if err = callback(records); err != nil {
		return tx.Abort(err)
	}
	return nil
}
func (r *PrivateRecords) check(key string) error {
	if r.closed {
		return groupstore.ErrTransaction
	}
	pattern := privateKeyPattern
	if r.namespace == "resource" {
		pattern = resourcePrivateKeyPattern
	} else if r.namespace == "contribution" {
		pattern = contributionPrivateKeyPattern
	}
	if !pattern.MatchString(key) {
		return privateIntegrity("chave de site inválida")
	}
	return nil
}
func (r *PrivateRecords) domain() string {
	if r.namespace == "resource" {
		return resourcePrivateDomain
	}
	if r.namespace == "contribution" {
		return contributionPrivateDomain
	}
	return privateDomain
}
func (r *PrivateRecords) aad(key string) []byte {
	data, _ := core.Canonical([]any{r.domain(), r.owner, r.storeID, key})
	return data
}
func (r *PrivateRecords) gcm(salt []byte) (cipher.AEAD, error) {
	key, err := hkdf.Key(sha256.New, r.secret, salt, r.domain(), 32)
	if err != nil {
		return nil, err
	}
	defer clear(key)
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}
func (r *PrivateRecords) encrypt(key string, plain []byte) ([]byte, error) {
	salt, nonce := make([]byte, 32), make([]byte, 12)
	if _, err := rand.Read(salt); err != nil {
		return nil, err
	}
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	gcm, err := r.gcm(salt)
	if err != nil {
		return nil, err
	}
	sealed := gcm.Seal(nil, nonce, plain, r.aad(key))
	out := make([]byte, 0, len(plain)+privateOverhead)
	out = append(out, 1)
	out = append(out, salt...)
	out = append(out, nonce...)
	out = append(out, sealed[len(sealed)-16:]...)
	out = append(out, sealed[:len(sealed)-16]...)
	return out, nil
}
func (r *PrivateRecords) decrypt(key string, encoded []byte) ([]byte, error) {
	if len(encoded) < privateOverhead || encoded[0] != 1 {
		return nil, privateIntegrity("envelope privado inválido")
	}
	gcm, err := r.gcm(encoded[1:33])
	if err != nil {
		return nil, err
	}
	sealed := append(append([]byte{}, encoded[61:]...), encoded[45:61]...)
	plain, err := gcm.Open(nil, encoded[33:45], sealed, r.aad(key))
	if err != nil {
		return nil, privateIntegrity("autenticação privada de site")
	}
	return plain, nil
}
func privatePart(key string, n int) string { return fmt.Sprintf("%s:%02d", key, n) }
func privateInt(value any) (int, error) {
	n, ok := value.(json.Number)
	if !ok {
		return 0, privateIntegrity("número privado inválido")
	}
	i, err := n.Int64()
	if err != nil || i < 0 || i > PrivateBytes+privateOverhead {
		return 0, privateIntegrity("número privado fora dos limites")
	}
	return int(i), nil
}
func (r *PrivateRecords) Read(key string) (value any, found bool, err error) {
	defer func() {
		if err != nil {
			err = r.tx.Abort(err)
		}
	}()
	if err = r.check(key); err != nil {
		return nil, false, err
	}
	manifest, found, err := r.tx.Get(key)
	if err != nil {
		return nil, false, err
	}
	keys, err := r.tx.Keys(key + ":")
	if err != nil {
		return nil, false, err
	}
	if !found {
		if len(keys) != 0 {
			return nil, false, privateIntegrity("blocos de site sem índice")
		}
		return nil, false, nil
	}
	parsed, err := core.DecodeJSON(manifest, 1024)
	if err != nil {
		return nil, false, privateIntegrity("índice privado ilegível")
	}
	m, ok := parsed.(map[string]any)
	if !ok || len(m) != 4 || m["domain"] != r.domain() {
		return nil, false, privateIntegrity("índice privado inválido")
	}
	size, err := privateInt(m["bytes"])
	if err != nil {
		return nil, false, err
	}
	chunks, err := privateInt(m["chunks"])
	if err != nil {
		return nil, false, err
	}
	digest, ok := m["digest"].(string)
	encoded, err := core.Canonical(m)
	if err != nil || !bytes.Equal(manifest, encoded) || !ok || !core.ValidAddress(digest) || size < privateOverhead || size > PrivateBytes+privateOverhead || chunks != (size+PrivateChunkBytes-1)/PrivateChunkBytes || chunks > PrivateChunks || len(keys) != chunks {
		return nil, false, privateIntegrity("limites privados inválidos")
	}
	combined := make([]byte, 0, size)
	for i, k := range keys {
		if k != privatePart(key, i) {
			return nil, false, privateIntegrity("conjunto privado incompleto")
		}
		piece, exists, e := r.tx.Get(k)
		if e != nil {
			return nil, false, e
		}
		expected := min(PrivateChunkBytes, size-i*PrivateChunkBytes)
		if !exists || len(piece) != expected {
			return nil, false, privateIntegrity("bloco privado truncado")
		}
		combined = append(combined, piece...)
	}
	if core.Hash(combined) != digest {
		return nil, false, privateIntegrity("hash privado inválido")
	}
	plain, err := r.decrypt(key, combined)
	if err != nil {
		return nil, false, err
	}
	value, err = core.DecodeJSON(plain, PrivateBytes)
	if err != nil {
		return nil, false, privateIntegrity("dados privados ilegíveis")
	}
	encoded, err = core.Canonical(value)
	if err != nil || !bytes.Equal(encoded, plain) {
		return nil, false, privateIntegrity("dados privados não canónicos")
	}
	return value, true, nil
}
func (r *PrivateRecords) Write(key string, value any) (err error) {
	defer func() {
		if err != nil {
			err = r.tx.Abort(err)
		}
	}()
	if err = r.check(key); err != nil {
		return err
	}
	plain, err := core.Canonical(value)
	if err != nil {
		return err
	}
	if len(plain) > PrivateBytes {
		return groupstore.ErrCapacity
	}
	encrypted, err := r.encrypt(key, plain)
	if err != nil {
		return err
	}
	chunks := (len(encrypted) + PrivateChunkBytes - 1) / PrivateChunkBytes
	if chunks > PrivateChunks {
		return groupstore.ErrCapacity
	}
	if err = r.Remove(key); err != nil {
		return err
	}
	for i := 0; i < chunks; i++ {
		if err = r.tx.Put(privatePart(key, i), encrypted[i*PrivateChunkBytes:min(len(encrypted), (i+1)*PrivateChunkBytes)], groupstore.Data); err != nil {
			return err
		}
	}
	manifest, err := core.Canonical(map[string]any{"domain": r.domain(), "bytes": len(encrypted), "chunks": chunks, "digest": core.Hash(encrypted)})
	if err != nil {
		return err
	}
	return r.tx.Put(key, manifest, groupstore.Data)
}
func (r *PrivateRecords) Remove(key string) (err error) {
	defer func() {
		if err != nil {
			err = r.tx.Abort(err)
		}
	}()
	if err = r.check(key); err != nil {
		return err
	}
	keys, err := r.tx.Keys(key + ":")
	if err != nil {
		return err
	}
	for _, k := range keys {
		if err = r.tx.Delete(k); err != nil {
			return err
		}
	}
	return r.tx.Delete(key)
}
